Implement Save and Save As

TomasHubelbauer opened the pull request:
Hello, I'd like to start trying to contribute to this Windows API based frontend. I am a novice to both Rust and Windows API (I do have experience with interoping with it from .NET so it shouldn't be too bad), so if you prefer my experiments stick to my fork, I will totally understand that.

In this PR I have introduced a *Save* menu item. Currently it behaves as a *Save as…* because I was not able to make pulling out the `filename` from the `state` work yet. I will keep trying (and figure out how to use a debugger etc.), so I opened the PR already in case it would be okay to do that in two PRs, otherwise I'll push the fix to this PR as well.

TomasHubelbauer commented on Cargo.lock:
```diff
@@ -1,22 +1,3 @@
-[root]
```
I am not sure why Cargo moved this. :-(

raphlinus replied:
I don't worry too much about Cargo changes.

TomasHubelbauer commented on src/dialog.rs:
```diff
@@ -0,0 +1,46 @@
+  get_file_dialog_path(hwnd_owner, false)
+}
+
+unsafe fn get_file_dialog_path(hwnd_owner: HWND, open: bool) -> Option<String> {
```
I decided to use a `bool` here even though it's poor modeling because I need to pass two pieces of info - the class and the ID, the they are both either `Open` or `Save` so I wanted to prevent an invalid invariant. This is then a two-state value but I felt creating an `enum` for this might have been an overkill given the two call sites in the same file? Not sure.

raphlinus replied:
It's reasonable.

TomasHubelbauer commented on src/main.rs:
```diff
@@ -89,6 +89,7 @@ struct MainWinState {
     dwrite_factory: directwrite::Factory,
     render_target: Option<RenderTarget>,
     resources: Option<Resources>,
+    filename: Option<String>,
```
Maybe it doesn't make sense to store the `filename` here but instead should be pulled in from the backend based on the view ID in the save operation? However it could also be used for window/tab title and I in that case maybe it's better to keep it locally as opposed to talking to the backend for it on each file / tab change? (Not that there's any tabs yet…). Dunno.

raphlinus replied:
There's definitely going to need to be a per-tab state object. This seems like as good a place as any for now though.

Basing it on the view id doesn't seem like a good idea. In the mac front-end, the filename is also managed by the front-end (actually as a url in the NSDocument infrastructure).

TomasHubelbauer commented on src/main.rs:
```diff
@@ -194,36 +196,31 @@ impl MainWin {
-            (*pfd).Release();
+        let filename = unsafe { get_open_file_dialog_path(hwnd_owner) };
+        if let Some(filename) = filename {
+            // TODO: Do not save here and pull from Xi core based on `self.state.view_id` instead?
```
See my review comment in `struct MainWinState` above, I will remove this line in the next commit.

TomasHubelbauer commented on src/main.rs:
```diff
@@ -194,36 +196,31 @@ impl MainWin {
+            self.state.borrow_mut().filename = Some(filename.clone());
+            // Note: this whole protocol has changed a lot since the
+            // 0.2 version of xi-core.
+            self.send_edit_cmd("open", &json!({
```
I intend to try and update `xi-core` and the commands used but I decided it would be best to do that in a separate PR.

raphlinus replied:
Sounds like a good idea. I can do a release of xi core if that would be helpful; it's been too long since the last one.

TomasHubelbauer commented on src/main.rs:
```diff
@@ -194,36 +196,31 @@ impl MainWin {
+        // let filename = &self.state.borrow().filename;
+        // if let &Some(ref filename) = filename {
+        //     self.send_edit_cmd("save", &json!({
+        //         "filename": filename,
```
I didn't know how to get around the `&Option` here and I could use help in figuring out how to make this condition work as expected.

raphlinus replied:
I don't see anything obviously wrong in the code, what exactly is the problem? I should probably patch this and play with it...

TomasHubelbauer commented:
Thanks for the review @raphlinus. I'll take a fresh look at the problem I had with the option fallback and post the error I'm getting here if I can't fix it. Please hold off on merging for now, I think I'll be able to resolve it.

TomasHubelbauer commented:
Hey @raphlinus. I made some further changes, but it will still need more work. I've split the buttons into *Save* and *Save As…*. I implemented the correct behavior of either saving or doing *Save As…* if no file is open, but saving currently crashes. The JSON looks fine, I debugged it with `MessageBox`, but clearly there must be some difference between the `filename` as seen in `file_save` and `file_save_as`, because it only crashes with `file_save` and the JSON is otherwise identical.

I [set upMSVC debugging](http://www.brycevandyk.com/debug-rust-on-windows-with-visual-studio-code-and-the-msvc-debugger/), so now I am able to visualize some values, but it didn't really help with zeroing in on the issue. I'll keep digging to see if I can find the problem and fix it.

TomasHubelbauer commented:
Hey @raphlinus, I've split the buttons into *Save* and *Save As…*. I implemented the correct behavior for *Save* of either saving directly or doing *Save As…* if no file is open.

I [set upMSVC debugging](http://www.brycevandyk.com/debug-rust-on-windows-with-visual-studio-code-and-the-msvc-debugger/) and contributed the debugger setup for VS Code as well. Let me know if you want me to move that to a separate PR perhaps.

That said, this PR should be now good to review as final.

TomasHubelbauer commented:
@raphlinus Do you think this can be merged?

raphlinus commented:
Oops, sorry about that, and thanks for the contribution!
